cdap-ui/app/services/mypromise.js

```diff
@@ -0,0 +1,183 @@
+/*
+  Purpose:
+    TL;DR
+    MyPromise is observable promise. Sounds very disturbing but this is the initial
+    attempt at sockets + $resource in an angular app.
+
+    Longer Version:
+    We cannot use promise pattern in a socket environment as promises
+    resolve only once. In the case of sockets we might want to 'poll' for a data
+    and get updated as soon something has changed.
+    MyPromise provides an interface simillar to a Promise (not $q) and accepts
+    in addition to a function a second argument. If you create your promise to
+    be an observable then the handlers are never erased and your callback/resolve
+    handler will be called whenever the promise gets resolved.
+
+    @param {function} which gets a 'resolve' and a 'reject' methods
+    @param {boolean} is observable or not.
+
+    PS: Inspired from
+      - https://www.promisejs.org/implementing/
+      - https://github.com/kriskowal/q/blob/v1/design/README.js
+*/
+angular.module(PKG.name + '.services')
+  .provider('MyPromise', function() {
+    var PENDING = 0;
```

Comment:
Not sure if I see the advantage here. Any specific reason it needs to be inside an object?